src/vm/checker/typecheck/tests/assets.rs

```diff
@@ -0,0 +1,173 @@
+use vm::types::AtomTypeIdentifier;
+use vm::parser::parse;
+use vm::checker::errors::CheckErrors;
+use vm::checker::{AnalysisDatabase, AnalysisDatabaseConnection};
+
+const FIRST_CLASS_TOKENS: &str = "(define-token stackaroos)
+         (define-asset stacka-nfts (buff 10))
+         (get-owner stacka-nfts \"1234567890\" )
+         (define-read-only (my-get-balance (account principal))
+            (get-balance stackaroos account))
+         (define-public (my-token-transfer (to principal) (amount int))
+            (transfer-token! stackaroos amount tx-sender to))
+         (define-public (faucet)
+           (let ((original-sender tx-sender))
+             (as-contract (transfer-token! stackaroos 1 tx-sender original-sender))))
+         (define-public (mint-after (block-to-release int))
+           (if (>= block-height block-to-release)
+               (faucet)
+               (err 8)))
+         (begin (mint-token! stackaroos 10000 'SZ2J6ZY48GV1EZ5V2V5RB9MP66SW86PYKKQ9H6DPR)
+                (mint-token! stackaroos 200 'SM2J6ZY48GV1EZ5V2V5RB9MP66SW86PYKKQVX8X0G)
+                (mint-token! stackaroos 4   'CTtokens))";
+
+const ASSET_NAMES: &str =
+        "(define burn-address 'SP000000000000000000002Q6VF78)
+         (define (price-function (name int))
+           (if (< name 100000) 1000 100))
+         
+         (define-asset names int)
```

Comment:
I like that we can now treat names as a first-class asset!  However, it occurs to me that when we produce the BNS contract, we will need to make `names` into `buff`s of various lengths (i.e. the type of a name would be `(tuple ((buff 37) (length int))`), such that the preorder hash of the name is only over the first `length` bytes of the name's `buff`.  I was wondering if it would be appropriate to try and evolve the `ASSET_NAMES` script into something closer to what the BNS contract will do, as a way to get a sanity check on the features of Clarity so far?  i.e. We can feasibly declare Clarity 1.0 once it's powerful enough to express BNS.

Reply:
Yeah -- that's pretty much where my thinking was: BNS would be implemented using assets, and it would give us a good sanity check on our features. That'll be a separate PR though!